numcodecs/compat.py

```diff
@@ -27,80 +26,96 @@
     from functools import reduce
 
 
-def buffer_tobytes(v):
-    """Obtain a sequence of bytes for the memory buffer used by `v`."""
-    if isinstance(v, binary_type):
-        return v
-    elif isinstance(v, np.ndarray):
-        return v.tobytes(order='A')
-    elif PY2 and isinstance(v, array.array):  # pragma: py3 no cover
-        return v.tostring()
-    else:
-        return memoryview(v).tobytes()
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
 
-def buffer_copy(buf, out=None):
-    """Copy the contents of the memory buffer from `buf` to `out`."""
+def ensure_contiguous_ndarray(buf):
+    """TODO"""
 
-    if out is None:
-        # no-op
-        return buf
+    if not isinstance(buf, np.ndarray):
 
-    # handle ndarray destination
-    if isinstance(out, np.ndarray):
+        # make that we create an array from a memory buffer with no copy
 
-        # view source as destination dtype
-        if isinstance(buf, np.ndarray):
-            buf = buf.view(dtype=out.dtype).reshape(-1, order='A')
-        else:
-            buf = np.frombuffer(buf, dtype=out.dtype)
+        if PY2:  # pragma: py3 no cover
+            try:
+                mem = memoryview(buf)
+            except TypeError:
+                # on PY2 also check if object exports old-style buffer interface
+                mem = np.getbuffer(buf)
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+        else:  # pragma: py2 no cover
+            mem = memoryview(buf)
+
+        arr = np.array(mem, copy=False)
 
-        # copy via numpy
-        np.copyto(out, buf)
+        if PY2 and isinstance(buf, array.array):  # pragma: py3 no cover
+
+            # type information will not have been propagated via the old-style buffer
+            # interface, so we have to manually hack it after the fact
+            if buf.typecode == 'u':
+                t = 'U1'
```

Comment:
This one gets more complicated.

First off the unicode array type is [totally broken]( https://bugs.python.org/issue13072 ) and [deprecated]( https://github.com/python/cpython/commit/029780be09ab5483ac4562d795ea4b4b6d456f44 ). Second it maps to `Py_UNICODE`, which can be [build dependent]( https://bugs.python.org/issue13072#msg168005 ). On Windows this is UCS-2 because of [historic reasons]( https://en.wikipedia.org/wiki/Unicode_in_Microsoft_Windows ). On other platforms this is usually UCS-4 (with the exception of [Python 2.7 on macOS]( https://github.com/Homebrew/legacy-homebrew/pull/32368 ) mostly due to community inertia). Third `memoryview`s represent UCS-2 and UCS-4 as different types (i.e. checking `.format`) where the first is `"u"` and the second is `"w"`. Fourth NumPy only supports UCS-4 not UCS-2 as noted in [this issue]( https://github.com/numpy/numpy/issues/12369 ). Fortunately NumPy will give a better error message about this in NumPy 1.16. 

If we are to go down the road of supporting these text types (which also includes `"c"` on Python 2.7), my suggestion would be that we just cast them all to appropriate width unsigned integers, which interestingly has [come up before]( https://bugs.python.org/issue13072#msg167997 ). Possibly raising a warning along the way to point out these types will be dropped in Python 4. That way this data can always safely be represented by NumPy or anything else without needing to know anything about these deprecated or removed text types. Alternatively we could raise an error if users give us this sort of data and not support it.

Reply:
Argh, OK, now the code in #121 makes sense.

Reply:
That is expected. Windows is using UCS-2 and NumPy only supports UCS-4. Casting is pretty much the only way around this AFAIK.

Reply:
OK, now I completely understand why there was array casting before and after conversion to numpy array. Thanks!

Reply:
Yeah, this wound up being cleaner to handle in Cython interestingly. 😆

`memoryview`'s `cast` is not nearly as functional as `ndarray`'s `view`.